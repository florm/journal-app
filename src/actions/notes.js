import { db } from "../firebase/firebase-config"
import { loadNotes } from "../helpers/loadNotes"
import { types } from "../types/types"
import Swal from 'sweetalert2'
import { fileUpload } from "../helpers/fileUpload"

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid  = getState().auth.uid
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        
        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote)
        dispatch(activarNota(docRef.id, newNote))
        dispatch(addnewNote(docRef.id, newNote))

    } 

}

export const activarNota = (id, note) => {
    return {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
}

export const addnewNote = (id, note) => {
    return {
        type: types.notesAddNew,
        payload: {
            id,
            ...note
        }
    }
}

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const startSaveNote = (note) => {
    return async (dispatch,getState) => {
        const { uid } = getState().auth
        if(!note.url){
            delete note.url
        }
        const noteToFirestore = {...note}
        delete noteToFirestore.id
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
        dispatch(refreshNote(note.id, note))
        Swal.fire('Saved', note.title, 'success')
    }
}

export const refreshNote = (id, note) =>{
    return {
        type: types.notesUpdated,
        payload: {
            id,
            note
        }
    }
}

export const startUploading = (file)=>{
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes
        
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: ()=>{
                Swal.showLoading()
            }
            
        })

        const fileUrl = await fileUpload(file)
        // console.log(fileUrl)
        Swal.close()
        activeNote.url = fileUrl
        dispatch(startSaveNote(activeNote))

    }
}

export const startDelete = (id)=>{
    return async (dispatch, getState) =>{
        const uid = getState().auth.uid

        await db.doc(`${uid}/journal/notes/${id}`).delete()

        dispatch(deleteNote(id))

    }
}

export const deleteNote = (id) => {
    return {
        type:types.notesDelete,
        payload: id
    }
}

export const noteLogout = () =>{
    return {
        type: types.notesLogoutCleaning
    }
}