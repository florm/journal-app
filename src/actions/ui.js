import { types } from "../types/types"

export const setError = (msgError) => {
    return{
        type: types.uiSetError,
        payload: msgError
    }
}

export const removeError = () => {
    return{
        type: types.uiRemoveError,
        payload: null
    }
}

export const startLoading = () => {
    return{
        type: types.uiStartLoading
    }
}

export const finishLoading = () => {
    return{
        type: types.uiFinishLoading
    }
}