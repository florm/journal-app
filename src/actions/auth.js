import { types } from "../types/types"
import { db, firebase, googleAuthProvider } from '../firebase/firebase-config'
import { finishLoading, startLoading } from "./ui"
import Swal from 'sweetalert2'
import { noteLogout } from "./notes"


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading())
        
        //el return lo agregue para tener todas las acciones en el test, porque 
        //sino es un metodo que no devuelve nada y para el test el codigo de abajo no 
        //se ejecuta
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())

            }).catch(e => {
                console.log(e)
                dispatch(finishLoading())
                Swal.fire('Error', e.message, 'error')
            })
    }
}
export const startRegisterWithEmailPassword = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name
                })
                dispatch(login(user.uid, user.displayName))
            }).catch(e => {
                console.log(e)
                Swal.fire('Error', e.message, 'error')
            })
    }
}

export const startGooleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(userCredential => {
                const { uid, displayName } = userCredential.user
                dispatch(login(uid, displayName))
            })
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut()
        dispatch(logout())
        dispatch(noteLogout())
    }
}
export const logout = () => {
    return {
        type: types.logout
    }
}


