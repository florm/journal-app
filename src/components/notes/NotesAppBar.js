import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import moment from 'moment'
import { startSaveNote, startUploading } from '../../actions/notes'

export const NotesAppBar = () => {
	const { active: notaActiva } = useSelector(state => state.notes)
	const dateNote = moment(notaActiva.date)
	const dispatch = useDispatch()

	const handleSave = () => {
		dispatch(startSaveNote(notaActiva))
	}

	const handlePictureUpload = () => {
		document.querySelector('#fileSelector').click()
	}

	const handleFileChange = (e) => {
		const file = e.target.files[0]
		if(file){
			dispatch(startUploading(file))
		}
	}

	return (
		<div className='notes__appbar'>
			<span>{dateNote.format("MMM D YYYY")}</span>

			<input
				id='fileSelector'
				style={{ display: 'none' }}
				type="file"
				onChange={handleFileChange}
			/>
			<div>
				<button className='btn'
					onClick={handlePictureUpload}
				>
					Picture
				</button>

				<button
					className='btn'
					onClick={handleSave}
				>
					Save
				</button>
			</div>
		</div>
	)
}
