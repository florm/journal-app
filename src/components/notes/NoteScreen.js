import React, {useEffect, useRef} from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activarNota, startDelete } from '../../actions/notes'
import { useForm } from '../../hooks/useForm'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
	
	const dispatch = useDispatch()
	const { active: note } = useSelector(state => state.notes)

	const [formValues, handleInputChange, reset] = useForm(note)
	const { title, body } = formValues
	
	const activeId = useRef(note.id)
	useEffect(() => {
	  
		if(note.id !== activeId.current){
			reset(note)
			activeId.current = note.id
		}
	 
	},[note, reset] )

	
	useEffect(()=> {
		dispatch(activarNota(formValues.id, { ...formValues}))
	},[dispatch, formValues])
	
	const handleDelete = ()=>{
		console.log(activeId)
		dispatch(startDelete(activeId.current))
	}
	return (
		<div className='notes__main-content'>
			<NotesAppBar />
			<div className='notes__content'>
				<input type="text"
					placeholder='Some awesome title'
					className='notes__title-input'
					autoComplete='off'
					name='title'
					value={title}
					onChange={handleInputChange}
					
				/>

				<textarea
					placeholder="What happened today"
					className='notes__text-area'
					name='body'
					value={body}
					onChange={handleInputChange}
				></textarea>
				{
					(note.url) &&
					<div className='notes__images'>
						<img src={note.url} alt="paisaje" />

					</div>}

			</div>

			<button
				className='btn btn-danger'
				onClick={ handleDelete}
			>Delete</button>
		</div>
	)
}
