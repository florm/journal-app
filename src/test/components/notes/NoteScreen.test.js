import {mount} from 'enzyme'
import { Provider } from 'react-redux'
import { Sidebar } from '../../../components/journal/Sidebar'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { NoteScreen } from '../../../components/notes/NoteScreen'
import { activarNota } from '../../../actions/notes'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

jest.mock('../../../actions/notes', ()=>({
    activarNota: jest.fn()
}))


const initalState = {
    auth:{
        name: 'Flor'
    },
    notes:{
        active: {
            id: '1234',
            title: 'Hola',
            body: 'Mundo',
            date: 0
        },
        notes: []
    }
}

const store = mockStore(initalState)
store.dispatch = jest.fn()

describe('pruebas en el NoteScreen', () => {

    const wrapper = mount(
        <Provider store={store}>
            <NoteScreen/>
        </Provider>
    )

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()

    })

    test('cuando se modifica el valor de un input se dispara el activarNota', () => {
        wrapper.find('input[name="title"]').simulate('change', { target:{
            name: 'title',
            value: 'hola de nuevo'
        }})

        //expect(activarNota).toHaveBeenLastCalledWith({})
        
        expect(activarNota).toHaveBeenCalledWith(
            '1234',
            {
                body: 'Mundo',
                title: 'hola de nuevo',
                id: '1234',
                date: 0
            }
        )
    })

    
})