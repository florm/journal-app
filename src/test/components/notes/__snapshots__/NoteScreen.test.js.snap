// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el NoteScreen debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Dec 31 1969
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__text-area"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
